go/parquet/metadata/file.go

```diff
@@ -0,0 +1,480 @@
+		if !algo.Aad.SupplyAadPrefix {
+			signingAlgo.Aad.AadPrefix = algo.Aad.AadPrefix
+		}
+		signingAlgo.Algo = parquet.AesGcm
+		f.metadata.EncryptionAlgorithm = signingAlgo.ToThrift()
+		footerSigningMetadata := f.props.FileEncryptionProperties().FooterKeyMetadata()
+		if footerSigningMetadata != "" {
+			f.metadata.FooterSigningKeyMetadata = []byte(footerSigningMetadata)
+		}
+	}
+
+	f.metadata.Schema = schema.ToThrift(f.schema.Root())
+	f.metadata.KeyValueMetadata = f.kvmeta
+
+	out := &FileMetaData{
+		FileMetaData: f.metadata,
+		version:      NewAppVersion(f.metadata.GetCreatedBy()),
+	}
+	if err := out.initSchema(); err != nil {
+		return nil, err
+	}
+	out.initColumnOrders()
+
+	f.metadata = format.NewFileMetaData()
+	f.rowGroups = nil
+	return out, nil
+}
+
+// KeyValueMetadata is an alias for a slice of thrift keyvalue pairs
+type KeyValueMetadata []*format.KeyValue
+
+// NewKeyValueMetadata is equivalent to make(KeyValueMetadata, 0)
+func NewKeyValueMetadata() KeyValueMetadata {
+	return make(KeyValueMetadata, 0)
+}
+
+func (k *KeyValueMetadata) Append(key, value string) {
+	*k = append(*k, &format.KeyValue{Key: key, Value: &value})
+}
+
+func (k KeyValueMetadata) Len() int { return len(k) }
+
+// Equals compares all of the metadata keys and values to check they are equal
+func (k KeyValueMetadata) Equals(other KeyValueMetadata) bool {
+	return reflect.DeepEqual(k, other)
+}
+
+func (k KeyValueMetadata) Keys() (ret []string) {
+	ret = make([]string, len(k))
+	for idx, v := range k {
+		ret[idx] = v.GetKey()
+	}
+	return
+}
+
+func (k KeyValueMetadata) Values() (ret []string) {
+	ret = make([]string, len(k))
+	for idx, v := range k {
+		ret[idx] = v.GetValue()
+	}
+	return
+}
+
+func (k KeyValueMetadata) FindValue(key string) *string {
+	for _, v := range k {
+		if v.Key == key {
+			return v.Value
+		}
+	}
+	return nil
+}
+
+// FileMetaData is a proxy around the underlying thrift FileMetaData object
+// to make it easier to use and interact with.
+type FileMetaData struct {
+	*format.FileMetaData
+	version       *AppVersion
+	Schema        *schema.Schema
+	FileDecryptor encryption.FileDecryptor
+	metadataLen   int
+}
+
+// NewFileMetaData takes in the raw bytes of the serialized metadata to deserialize
+// and will attempt to decrypt the footer if a decryptor is provided.
+func NewFileMetaData(data []byte, fileDecryptor encryption.FileDecryptor) (*FileMetaData, error) {
+	meta := format.NewFileMetaData()
+	if fileDecryptor != nil {
+		footerDecryptor := fileDecryptor.GetFooterDecryptor()
+		data = footerDecryptor.Decrypt(data)
+	}
+
+	remain, err := thrift.DeserializeThrift(meta, data)
+	if err != nil {
+		return nil, err
+	}
+
+	f := &FileMetaData{
+		FileMetaData:  meta,
+		version:       NewAppVersion(meta.GetCreatedBy()),
+		metadataLen:   len(data) - int(remain),
+		FileDecryptor: fileDecryptor,
+	}
+
+	f.initSchema()
+	f.initColumnOrders()
+
+	// init keyvalue metadata
```

Comment:
Ah, that was a todo from an earlier version before i created that `KeyValueMetadata` type and the `NewKeyValueMetadata` function which initialized one. It can be removed, good catch.